# Obrada informacija

## Laboratorijska vježba 2

U ovoj vježbi upoznat ćete se s jednom primjenom tehnika obrade informacija u bioinformatici. Ova laboratorijska vježba nosi 4 boda. Izvješće s ove laboratorijske vježbe potrebno je predati u .pdf formatu na *Moodle*. Izvješće koje predajete se mora zvati *PrezimeIme.pdf*. 

Osim biblioteka za rad s Fourierovom transformacijom (koristit ćemo samo numpy) koristit ćemo i biblioteku biopython koja sadrži puno korisnih alata iz područja bioinformatike. Mi ćemo je koristiti za jednostavnije baratanje bioinformatičkim tipovima podataka.

Biblioteka biopython dolazi s instalacijom Anaconde, ali ju je potrebno uključiti u okolinu (*environment*) koja se koristi.

Ako vježbu izvodite u Google Colab okruženju, morate instalirati biblioteku biopython. Instalaciju je potrebno izvršiti u sklopu prvog zadatka ove laboratorijske vježbe. Instalaciju izvodite sljedećim kodom: 

    try:
      import google.colab
      !pip install biopython
    except ImportError:
      pass

Nakon izvođenja ovog koda, možete učitati biopython biblioteku. 

1. Zadatak

Python biblioteke potrebne za laboratorijske vježbe su numpy i biopython. Uključite ih ("importirajte") i ispišite verziju svake od njih pomoću \[ime_biblioteke\].__version__.

UPUTA: Osnovna biopython biblioteka ima naziv Bio.

In [1]:
import Bio
import numpy as np
print("Numpy version - " + np.__version__)
print("Bio version - " + Bio.__version__)

Numpy version - 1.26.4
Bio version - 1.84


2. Zadatak

Uz laboratorijske vježbe dobili ste dvije datoteke s podacima. Datoteku koja sadrži referentni genom jednog soja bakterije  Escherichia coli (escherichia_coli_reference.fasta) u FASTA formatu i datoteku koja sadrži skup očitanja dobivenih sekvenciranjem (ecoli_ILL_small.fastq) u FASTQ formatu.

Datoteke možete učitati koristeći metodu _parse()_ iz biblioteke Bio.SeqIO. Metoda _parse()_ vraća iterator koji možete pretvoriti u Python listu na sljedeći način:

    reads = list(parse("ime_datoteke", "tip_datoteke"))

Tip datoteke postavite na "fasta" ili "fastq".

Učitajte obje datoteke te ispišite broj zapisa u svakoj od njih (broj elemenata u listi). Datoteka koja sadrži referencu trebala bi imati samo jedan zapis, dok bi datoteka s očitanjima trebala sadržavati veći broj zapisa.

NAPOMENA: Ako niste sigurni kako pronaći datoteke na disku iz Jupyter notebook-a, uvijek možete provjeriti radni direktorij sljedećim naredbama:

    import os
    os.getcwd()

i promijeniti ga sa:

    os.chdir()

Ako pak radite u Google Colab okruženju, koristite upute za učitavanje datoteka s Google diska iz prve laboratorijske vježbe. 

In [2]:
from Bio import SeqIO
reads = list(SeqIO.parse("ecoli_ILL_small.fastq", "fastq"))
reference = list(SeqIO.parse("escherichia_coli_reference.fasta", "fasta"))
print("Broj zapisa u očitanjima - " + str(len(reads)))
print("Broj zapisa u referenci - " + str(len(reference)))

Broj zapisa u očitanjima - 38585
Broj zapisa u referenci - 1


3. Zadatak

Svaki zapis koji ste učitali pomoću metode _Bio.SeqIO.parse()_ sadrži Veći broj podataka od kojih su nam bitni samo neki. Naredbom print ispišite cijeli prvi zapis iz datoteke s očitanjima i iz datoteke s referencom.

Vidjet ćete da oba zapisa (među ostalim podacima) sadrže identifikator zapisa i sekvencu. Identifikator zapisa možete dohvatiti pomoću

    zapis.id

dok sekvencu možete dohvatiti pomoću

    zapis.seq

Ispišite identifikator i sekvencu za prvo očitanje te identifikator i prvih 200 znakova za referentni genom E.coli.

NAPOMENA: Referentni genom Escherichia coli je dugačak oko 4.5 milijuna slova

In [3]:
print("Prvi zapis iz ocitanja - " + reads[0])
print("Prvi zapis iz reference - " + reference[0])
print("Identifikator ocitanja - " + reads[0].id)
print("Sekvenca 1. ocitanja - " + reads[0].seq)
print("Identifikator reference - " + reference[0].id)
print("Prvih 200 znakova reference - " + reference[0].seq[0:200])

ID: SRR2052522.671
Name: SRR2052522.671
Description: SRR2052522.671 HWI-EAS390_0001:4:1:6915:1123/1
Number of features: 0
Seq('Prvi zapis iz ocitanja - GATCTGGTGACCGGGTCGCGCAAAGTGAT...TGC')
ID: NC_000913.3
Name: NC_000913.3
Description: NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome
Number of features: 0
Seq('Prvi zapis iz reference - AGCTTTTCATTCTGACTGCAACGGGCAA...TTC')
Identifikator ocitanja - SRR2052522.671
Sekvenca 1. ocitanja - GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTTGGGCCTCTGTATCATGCCACTCACTGCGCAATATCCGGATCAAATGC
Identifikator reference - NC_000913.3
Prvih 200 znakova reference - AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATATAGGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCATGAAACGCAT


4. Zadatak

Da bismo sekvence DNA analizirali metodama obrade signala, moramo pojednim nukleotidima (slovima) dodijeliti brojčane vrijednosti. Napišite funkciju u Pythonu koja će primiti slovo koje predstavlja nukleotid i vratiti odgovarajuću brojčanu vrijednost. Vrijednosti dodijelite na sljedeći način:

- A =  3
- G =  2
- C = -2
- T = -3

DNA sekvence mogu sadržavati i neke druge znakove (npr. 'N' koji označava da taj nuklotid nije poznat), njima dodijelite vrijednost 0. Također se može dogoditi da nukleotidi budu označeni i malim slovima, pa vodite računa da vaša funkcija mora vratiti ispravnu vrijendost i u tom slučaju.

In [4]:
def pretvoriUSignal(slovo):
    if slovo.upper() == "A":
        return 3
    elif slovo.upper() == "G":
        return 2
    elif slovo.upper() == "C":
        return -2
    elif slovo.upper() == "T":
        return -3
    else:
        return 0

5. Zadatak

Upotrebite napisanu funkciju da bi od prvog očitanja i od reference kreirali signal. Izračunajte korelaciju pomoću Fourierove transformacije. Zanemarite imaginarne vrijednosti.

In [5]:
referenca_signal = [pretvoriUSignal(s) for s in reference[0].seq]
ocitanje_signal = [pretvoriUSignal(s) for s in reads[0].seq]

k_arr = range(-len(ocitanje_signal)+1, len(referenca_signal))
std = np.std([3, 2, -2, -3, 0])
#avg ne racunam jer je 0
referenca_signal = [x/std for x in referenca_signal]
ocitanje_signal = [x/std for x in ocitanje_signal]

padding1 = [0]*(len(ocitanje_signal)-1)
padding2 = [0]*(len(referenca_signal)-1)

X1 = np.fft.fft(padding1 + referenca_signal)
X2 = np.fft.fft(ocitanje_signal + padding2)
Cor = np.conjugate(X2)*X1
cor = np.fft.ifft(Cor)
cor_real = np.real(cor)
k = k_arr[cor.argmax()]
print("Korelacija po FFT-u:")
print(cor_real)
print("k = " + str(k))

Korelacija po FFT-u:
[-1.15384615  0.38461538 -0.19230769 ... -1.73076923 -2.30769231
 -0.76923077]
k = 2324486


6. Zadatak

Ispišite duljinu reference. Koristeći metode biblioteke _numpy_, izračunajte srednju vrijednost i standardnu devijaciju duljine očitanja (uzmite u obzir sva očitanja).

Primijetit ćete da su sva očitanja jednake duljine.

In [6]:
print("Duljina reference - " + str(len(reference[0].seq)))
duljine_ocitanja = [len(read.seq) for read in reads]
print("Srednja vrijednost duljine ocitanja - " + str(np.array(duljine_ocitanja).mean()))
print("Standardna devijacija duljine ocitanja - " + str(np.array(duljine_ocitanja).std()))

Duljina reference - 4641652
Srednja vrijednost duljine ocitanja - 121.0
Standardna devijacija duljine ocitanja - 0.0


7. zadatak

Što ako želimo izračunati korelaciju za veći broj očitanja i istu referencu? To je tipičan slučaj u bioinformatici jer uređaji za sekvenciranje proizvode tisuće i desetke tisuća očitanja koja se potom mapiraju na istu referencu.

Ako korelaciju računamo izravno, potrebno ju je svaki put izračunati iz početka. Ako korelaciju računamo pomoću FFT-a, transformaciju za referencu potrebno je napraviti samo jednom.

Izračunajte korelacije za prvih 10 očitanja.

In [7]:
for i in range(10):
    referenca_signal = [pretvoriUSignal(s) for s in reference[0].seq]
    ocitanje_signal = [pretvoriUSignal(s) for s in reads[i].seq]

    k_arr = range(-len(ocitanje_signal)+1, len(referenca_signal))
    std = np.std([3, 2, -2, -3, 0])
    #avg ne racunam jer je 0
    referenca_signal = [x/std for x in referenca_signal]
    ocitanje_signal = [x/std for x in ocitanje_signal]

    padding1 = [0]*(len(ocitanje_signal)-1)
    padding2 = [0]*(len(referenca_signal)-1)

    X1 = np.fft.fft(padding1 + referenca_signal)
    X2 = np.fft.fft(ocitanje_signal + padding2)
    Cor = np.conjugate(X2)*X1
    cor = np.fft.ifft(Cor)
    cor_real = np.real(cor)
    k = k_arr[cor_real.argmax()]
    print("Korelacija po FFT-u:")
    print(cor_real)
    print("k = " + str(k) + "\n")

Korelacija po FFT-u:
[-1.15384615  0.38461538 -0.19230769 ... -1.73076923 -2.30769231
 -0.76923077]
k = 2324486

Korelacija po FFT-u:
[ 1.73076923e+00  4.88497530e-15 -3.65384615e+00 ... -1.73076923e+00
 -2.50000000e+00 -1.15384615e+00]
k = 1877260

Korelacija po FFT-u:
[-1.15384615 -2.5        -2.11538462 ... -3.46153846 -1.92307692
 -0.76923077]
k = 557777

Korelacija po FFT-u:
[ 1.73076923 -0.57692308 -1.15384615 ... -1.73076923 -0.96153846
 -1.15384615]
k = 1144877

Korelacija po FFT-u:
[ 1.73076923e+00 -1.86517420e-14 -3.07692308e+00 ...  1.92307692e-01
 -9.61538462e-01 -1.15384615e+00]
k = 3583639

Korelacija po FFT-u:
[ 1.15384615  1.92307692  1.73076923 ... -4.03846154 -2.5
 -1.15384615]
k = 4051518

Korelacija po FFT-u:
[-1.15384615e+00  9.61538462e-01  3.07692308e+00 ... -1.73076923e+00
  5.32907172e-15  7.69230769e-01]
k = 2293706

Korelacija po FFT-u:
[ 1.15384615  2.5        -1.34615385 ...  1.73076923  0.96153846
  1.15384615]
k = 1011323

Korelacija po FFT-u:
[-1.7307692

8. zadatak

Na temelju najveće vrijednosti korelacije između reference i prvog očitanja pronađite poziciju na referenci koja je najsličnija očitanju. Pozicija odgovara vrijednosti parametra k za koji je korelacija najveća.

Napišite metodu koja će primiti dva niza znakova jednake duljine, usporediti znakove na istim pozicijama i vratiti broj razlika (Hammingova udaljenost).

"Izrežite" dio reference koji je najsličniji očitanju (iste duljine kao i očitanje) i usporedite ga s očitanjem pomoću napisane funkcije.

In [8]:
def hammingUdaljenost(niz1, niz2):
    rez = 0
    if len(niz1)==len(niz2):
        for i in range(len(niz1)):
            if niz1[i]!=niz2[i]:
                rez+=1
    else:
        raise ValueError("Nizovi moraju biti iste duljine")
    return rez;

k = 2324486 # za ovaj k je korelacija najveca
referenca_dio = reference[0].seq[k:k+len(reads[0].seq)]
print("Hammingova udaljenost - " + str(hammingUdaljenost(reads[0].seq, referenca_dio)))

Hammingova udaljenost - 9


9. zadatak

U datoteci "ecoli_ILL_small_aln.sam" dana su već izračunata poravnanja svih očitanja na referencu u SAM formatu. SAM je tekstualni "tab separated" format. U prvom stupcu se nalati identifikator očitanja, dok se u četvrtom stupcu nalazi pozicija na referenci na koju je očitanje najbolje poravnato (ostali stupci nas ne zanimaju). Također, datoteka s poravnanjima sadrži i nekoliko _header_ readaka kojima prvi stupac počinje sa znakom '@', njih također možete zanemariti.

Otvorite datoteku s poravnanjima i pronađite poravnanje za prvo očitanje (identifikator očitanja u datoteci s očitanjima i datoteci s poravnanjima mora biti jednak). Usporedite poziciju u datoteci sa pozicijom koju ste dobili pomoću korelacije.

UPOUTA: TSV datoteke možete otvoriti na sljedeći način:

    tsv_file = open("file_name")
    tsv_rows = csv.reader(tsv_file, delimiter="\t")

Varijabla _tsv\_rows_ će sadržavati listu redaka, a svaki redak biti lista vrijednosti (po jedna za svaki stupac).

In [9]:
import csv
tsv_file = open("ecoli_ILL_small_aln.sam")
tsv_rows = csv.reader(tsv_file, delimiter="\t")
for row in tsv_rows:
    if row[0] == reads[0].id:
        print("Poravnanje - " + row[3])

Poravnanje - 2324487


10. zadatak

Za prvo očitanje pozicija dobivena pomoću korelacije trabala bi biti 2324486, dok je pozicija u datoteci s poravnanjima 2324487. Razlikuju se samo za 1 pa možemo zaključiti da nam je korelacija dala dobru poziciju za poravnanje.

Prisjetimo se da korelacija ne računa točno poravnanje već ju koristimo samo da bi našli kandidatne pozicije za točno računanje. Tek onda na takvim pozicijama možemo točno poravnanje izračunati pomoću algoritama dinamičkog programiranja. Ako bi primijenili dinamičko programiranje za računanje poravnanja očitanja s cijelom referencom, postupak bi bio znatno sporiji i zahtijevao bi veliku količinu radne memorije.

**Ako želite** to možete isprobati pomoću algoritama za poravnanje u biblioteci _bioparser_. Lokalno poravnanje možete izračunati na sljedeći način:

    from Bio import Align
    aligner = Align.PairwiseAligner()
    aligner.mode = "local"
    alignments = aligner.align(reference[0].seq[pos-20:pos+len(read_sig)+20], reads[1].seq)

Za **drugu** skupinu od 100 očitanja (101. - 200.) izračunajte korelaciju te pomoću korelacije poziciju najveće sličnosti očitanja i reference. Usporedite rezultat sa podacima u datoteci s poravnanjima. Ispišite broj očitanja za koja se dvije pozicije razlikuju za najviše 5 mjesta.

In [12]:
counter = 0;
import csv

for i in range(100):
    referenca_signal = [pretvoriUSignal(s) for s in reference[0].seq]
    ocitanje_signal = [pretvoriUSignal(s) for s in reads[100+i].seq]

    k_arr = range(-len(ocitanje_signal)+1, len(referenca_signal))
    std = np.std([3, 2, -2, -3, 0])
    #avg ne racunam jer je 0
    referenca_signal = [x/std for x in referenca_signal]
    ocitanje_signal = [x/std for x in ocitanje_signal]

    padding1 = [0]*(len(ocitanje_signal)-1)
    padding2 = [0]*(len(referenca_signal)-1)

    X1 = np.fft.fft(padding1 + referenca_signal)
    X2 = np.fft.fft(ocitanje_signal + padding2)
    Cor = np.conjugate(X2)*X1
    cor = np.fft.ifft(Cor)
    cor_real = np.real(cor)
    k = k_arr[cor_real.argmax()]
    tsv_file = open("ecoli_ILL_small_aln.sam")
    tsv_rows = csv.reader(tsv_file, delimiter="\t")
    for row in tsv_rows:
        if row[0] == reads[100+i].id:
            if abs(int(row[3])-k)<=5:
                counter+=1

print("Broj ocitanja za koje se dvije pozicije razlikuju za najvise 5 mjesta - " + str(counter))

Broj ocitanja za koje se dvije pozicije razlikuju za najvise 5 mjesta - 41


11. zadatak

Očekivani broj točno pozicioniranih očitanja je otprilike 50, jer smo za sada uspješno radili samo s očitanjima na jednom lancu DNA.

Odaberite jedno očitanje koje NIJE uspješno pozicionirano. Ono bi se trebalo nalaziti na drugom (reverzno-kompleentarnom) DNA lancu. Napravite reverzni komplement odabranog očitanja. Možete koristiti metodu reverse_complement klase Seq.

    rc_seq = seq.reverse_complement()

Izračunajte pomoću korelacije poziciju poravnanja za reverzno komplementarno očitanje, te je usporedite s pozicijom iz datoteke. Sada bi se trebale poklapati!

In [10]:
rc_seq = reads[155].seq.reverse_complement()
referenca_signal = [pretvoriUSignal(s) for s in reference[0].seq]
ocitanje_signal = [pretvoriUSignal(s) for s in rc_seq]

k_arr = range(-len(ocitanje_signal)+1, len(referenca_signal))
std = np.std([3, 2, -2, -3, 0])
#avg ne racunam jer je 0
referenca_signal = [x/std for x in referenca_signal]
ocitanje_signal = [x/std for x in ocitanje_signal]

padding1 = [0]*(len(ocitanje_signal)-1)
padding2 = [0]*(len(referenca_signal)-1)

X1 = np.fft.fft(padding1 + referenca_signal)
X2 = np.fft.fft(ocitanje_signal + padding2)
Cor = np.conjugate(X2)*X1
cor = np.fft.ifft(Cor)
cor_real = np.real(cor)
k = k_arr[cor_real.argmax()]
print("k = " + str(k))
tsv_file = open("ecoli_ILL_small_aln.sam")
tsv_rows = csv.reader(tsv_file, delimiter="\t")
for row in tsv_rows:
    if row[0] == reads[155].id:
        print("Poravnanje - " + row[3])

k = 765096
Poravnanje - 765097


12. ZAKLJUČAK



Prolaskom kroz zadatke u ovoj vježbi dobili ste kratak uvod u rad s bioinformatičkim podacima i tehnikama obrade signala.